acp/identity/identity.go

```diff
@@ -34,33 +42,95 @@ type Identity struct {
 	// The address is derived from the actor's public key,
 	// using the did:key method
 	DID string
+
+	BearerToken string
 }
 
 // FromPrivateKey returns a new identity using the given private key.
-func FromPrivateKey(privateKey *secp256k1.PrivateKey) (immutable.Option[Identity], error) {
-	pubKey := privateKey.PubKey()
-	did, err := DIDFromPublicKey(pubKey)
+//
+//   - duration: The [time.Duration] that this identity is valid for.
+//   - audience: The audience that this identity is valid for.  This is required
+//     by the Defra http client.  For example `github.com/sourcenetwork/defradb`
+//   - authorizedAccount: An account that this identity is authorizing to make
+//     SourceHub calls on behalf of this actor.  This is currently required when
+//     using the SourceHub ACP module.
+func FromPrivateKey(
+	privateKey *secp256k1.PrivateKey,
+	duration time.Duration,
+	audience immutable.Option[string],
+	authorizedAccount immutable.Option[string],
+) (Identity, error) {
+	publicKey := privateKey.PubKey()
+	did, err := DIDFromPublicKey(publicKey)
+	if err != nil {
+		return Identity{}, err
+	}
+
+	subject := hex.EncodeToString(publicKey.SerializeCompressed())
+	now := time.Now()
+
+	jwtBuilder := jwt.NewBuilder()
+	jwtBuilder = jwtBuilder.Subject(subject)
+	jwtBuilder = jwtBuilder.Expiration(now.Add(duration))
+	jwtBuilder = jwtBuilder.NotBefore(now)
+	jwtBuilder = jwtBuilder.Issuer(did)
+	jwtBuilder = jwtBuilder.IssuedAt(now)
+
+	if audience.HasValue() {
+		jwtBuilder = jwtBuilder.Audience([]string{audience.Value()})
+	}
+
+	token, err := jwtBuilder.Build()
+	if err != nil {
+		return Identity{}, err
+	}
+
+	if authorizedAccount.HasValue() {
+		err = token.Set(acptypes.AuthorizedAccountClaim, authorizedAccount.Value())
+		if err != nil {
+			return Identity{}, err
+		}
+	}
+
+	signedToken, err := jwt.Sign(token, jwt.WithKey(AuthTokenSignatureScheme, privateKey.ToECDSA()))
 	if err != nil {
-		return None, err
+		return Identity{}, err
 	}
 
-	return immutable.Some(Identity{
-		DID:        did,
-		PublicKey:  pubKey,
-		PrivateKey: privateKey,
-	}), nil
+	return Identity{
+		DID:         did,
+		PrivateKey:  privateKey,
+		PublicKey:   publicKey,
+		BearerToken: string(signedToken),
+	}, nil
 }
 
-// FromPublicKey returns a new identity using the given public key.
-func FromPublicKey(publicKey *secp256k1.PublicKey) (immutable.Option[Identity], error) {
-	did, err := DIDFromPublicKey(publicKey)
+func FromToken(data []byte) (Identity, error) {
+	token, err := jwt.Parse(data, jwt.WithVerify(false))
```

Comment:
question: not familiar with the flow, will the token signature always be verified elsewhere?

Reply:
Yes, the Defra http client will verify this, and so will SourceHub.

Reply:
:saluting_face: